src/java.base/share/classes/java/util/concurrent/ForkJoinPool.java

```diff
@@ -2760,49 +2757,93 @@ static int getSurplusQueuedTaskCount() {
+            now = false;
+        else if ((e & STOP) != 0L)
+            now = true;
+        else if (now) {
+            if (((ps = getAndBitwiseOrRunState(SHUTDOWN|STOP) & STOP)) == 0L) {
+                if ((ps & RS_LOCK) != 0L) {
+                    spinLockRunState(); // ensure queues array stable after stop
+                    unlockRunState();
+                }
                 interruptAll();
+            }
         }
-        if ((e & (STOP | TERMINATED)) == STOP) { // help cancel tasks
-            if ((ctl & RC_MASK) > 0L) {          // unless all inactive
-                int r = (int)Thread.currentThread().threadId();
-                WorkQueue[] qs = queues;         // stagger traversals
-                int n = (qs == null) ? 0 : qs.length;
-                for (int l = n; l > 0; --l, ++r) {
-                    WorkQueue q; ForkJoinTask<?> t;
-                    if ((q = qs[r & (n - 1)]) != null &&
-                        q.source != DROPPED) {
-                        while ((t = q.poll()) != null) {
-                            try {
-                                t.cancel(false);
-                            } catch (Throwable ignore) {
-                            }
-                        }
+        else if ((isShutdown = (e & SHUTDOWN)) != 0L || enable) {
+            if (isShutdown == 0L)
+                getAndBitwiseOrRunState(SHUTDOWN);
+            if (quiescent() > 0)
+                now = true;
+        }
+        if (now) {
+            releaseWaiters();
+            for (;;) {
+                if (((e = runState) & CLEANED) == 0L) {
+                    boolean clean = cleanQueues();
+                    if (((e = runState) & CLEANED) == 0L && clean)
+                        e = getAndBitwiseOrRunState(CLEANED) | CLEANED;
+                }
+                if ((e & TERMINATED) != 0L)
+                    break;
+                if (ctl != 0L) // else loop if didn't finish cleaning
+                    break;
+                if ((e & CLEANED) != 0L) {
+                    e |= TERMINATED;
+                    if ((getAndBitwiseOrRunState(TERMINATED) & TERMINATED) == 0L) {
+                        CountDownLatch done; SharedThreadContainer ctr;
+                        if ((done = termination) != null)
+                            done.countDown();
+                        if ((ctr = container) != null)
+                            ctr.close();
                     }
+                    break;
                 }
             }
-            if (((e = runState) & TERMINATED) == 0L && ctl == 0L) {
-                e |= TERMINATED;
-                if ((getAndBitwiseOrRunState(TERMINATED) & TERMINATED) == 0L) {
-                    CountDownLatch done; SharedThreadContainer ctr;
-                    if ((done = termination) != null)
-                        done.countDown();
-                    if ((ctr = container) != null)
-                        ctr.close();
+        }
+        return e;
```

Comment:
@DougLea Is there any benefit to using `break` in the loop rather than `return e;` directly?

Reply:
It's marginal, but the current structure has a single return vs many that could have been done here, leading to simpler flow graph for compiler.. (In some other methods, it's not worth doing.)